core/file.go

```diff
@@ -156,57 +164,296 @@ func (file *File) Evaluate(ctx context.Context) (*buildkit.Result, error) {
+	}
+	n, err := cw.Writer.Write(p)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get buildkit client: %w", err)
+		return n, err
 	}
+	cw.wrote += n
+	return n, nil
+}
 
-	ref, err := bkRef(ctx, bk, file.LLB)
+// Reader returns an io.Reader for the file contents
+func (file *File) ReadCloser(ctx context.Context) (io.ReadCloser, error) {
+	ref, err := getRefOrEvaluate(ctx, file)
+	if err != nil {
+		return nil, err
+	}
+
+	mountable, err := ref.Mount(ctx, true, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get mountable reference: %w", err)
+	}
+
+	lm := snapshot.LocalMounter(mountable)
+	defer lm.Unmount()
+
+	dir, err := lm.Mount()
+	if err != nil {
+		return nil, err
+	}
+
+	resolvedPath, err := containerdfs.RootPath(dir, file.File)
 	if err != nil {
 		return nil, err
 	}
+	f, err := os.Open(resolvedPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file %q: %w", file.File, err)
+	}
+	return &fileReader{
+		mount: lm,
+		f:     f,
+	}, nil
+}
+
+type fileReader struct {
+	mount snapshot.Mounter
+	f     *os.File
+}
+
+func (r *fileReader) Read(p []byte) (n int, err error) {
+	return r.f.Read(p)
+}
 
-	// Stat the file and preallocate file contents buffer:
-	st, err := file.Stat(ctx)
+func (r *fileReader) Close() error {
+	return errors.Join(
+		r.f.Close(),
+		r.mount.Unmount(),
+	)
+}
+
+func (file *File) Search(ctx context.Context, opts SearchOpts) ([]*SearchResult, error) {
+	ref, err := getRefOrEvaluate(ctx, file)
 	if err != nil {
 		return nil, err
 	}
+	if ref == nil {
+		// empty directory, i.e. llb.Scratch()
+		return []*SearchResult{}, nil
+	}
 
-	// Error on files that exceed MaxFileContentsSize:
-	fileSize := int(st.GetSize_())
-	if fileSize > buildkit.MaxFileContentsSize {
-		// TODO: move to proper error structure
-		return nil, fmt.Errorf("file size %d exceeds limit %d", fileSize, buildkit.MaxFileContentsSize)
+	opt, ok := buildkit.CurrentOpOpts(ctx)
+	if !ok {
+		return nil, fmt.Errorf("no buildkit opts in context")
 	}
 
-	// Allocate buffer with the given file size:
-	contents := make([]byte, fileSize)
+	ctx = trace.ContextWithSpanContext(ctx, opt.CauseCtx)
 
-	// Use a chunked reader to overcome issues when
-	// the input file exceeds MaxFileContentsChunkSize:
-	var offset int
-	for offset < fileSize {
-		chunk, err := ref.ReadFile(ctx, bkgw.ReadRequest{
-			Filename: file.File,
-			Range: &bkgw.FileRange{
-				Offset: offset,
-				Length: buildkit.MaxFileContentsChunkSize,
-			},
-		})
+	bkSessionGroup, ok := buildkit.CurrentBuildkitSessionGroup(ctx)
+	if !ok {
+		return nil, fmt.Errorf("no buildkit session group in context")
+	}
+
+	results := []*SearchResult{}
+	err = MountRef(ctx, ref, bkSessionGroup, func(root string) error {
+		resolvedDir, err := containerdfs.RootPath(root, filepath.Dir(file.File))
+		if err != nil {
+			return err
+		}
+		rgArgs := opts.RipgrepArgs()
+		rgArgs = append(rgArgs, filepath.Base(file.File))
```

Comment:
We probably want a `--` before the path here, right?

Also, a check that file.File isn't absolute is probably worth it; there's been some confusion in the past around `Directory.Dir` and `File.File` being absolute vs. relative, using an abs path here accidentally would be not-good.

Reply:
It's already wrapped in a `filepath.Base` (with `filepath.Dir` passed to `containerdfs.RootPath` above), so I _think_ it should be fine when it comes to absolute vs. relative, but I've add a `--` to handle any funky filenames.